### PR TITLE
adjust the database, new admin-controller and change the startpage

### PR DESCRIPTION
**Pull request by tmshl**
- naming in database normalized
- new table device
- new table room mow with names
- table message_type now with translated message types

- controller admin with action reset to create new tables and test to fill with test data
- controller index changed to show a more minimalized startpage

### CODE REVIEW

**Comment by tmshl** [comment]
Please check the changes. Feel free to merge this branch into master if everything is ok.

**Comment by tmshl** [comment]
There is no new database-dump included. Just call /admin/reset and the tables are created automatically. /admin/test fills them with some values.

File: application/lib/DeviceDbFactory.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+		$device = new Device();
+
+      if (!isset($row["{$prefix}id"])) {
+         $row["{$prefix}id"] = 0;
+      }
+      else {
+         $row["{$prefix}id"] = intval($row["{$prefix}id"]);
+      }
+      
+      $this->fill($device, $row, $prefix);
```
Comment (douira) [inline application/lib/DeviceDbFactory.php]: what exactly does fill do?

File: application/lib/MessageDb.php [hunk 2]
```diff
@@ -1,324 +1,128 @@
...
+            .(count($bugStatus) > 0 ? " AND bug_status_id IN ($bugStatusId)" : "")
+      );
+      $query->bindValue(":message_type_id", $messageType->id, PDO::PARAM_INT);
+      $query->execute();
+      
+      $result = $query->fetch(PDO::FETCH_NUM);
+      return $result[0];
+   }
+   
    public function getAll() {
```
Comment (douira) [inline application/lib/MessageDb.php]: what does getAll do? get all messages?

File: application/lib/MessageDbFactory.php [hunk 3]
```diff
@@ -4,9 +4,9 @@
 class MessageDbFactory extends \MVC\db\DbFactory {
```
Comment (douira) [inline application/lib/MessageDbFactory.php]: is this class used with ::getInstance() ?

File: application/lib/MessageDbFactory.php [hunk 4]
```diff
@@ -27,7 +27,7 @@ public function create($row=array(), $prefix='') {
       $message->type  = $this->types[intval($row[$prefix."typeId"])];
       $message->module = $this->modules[intval($row[$prefix."moduleId"])];
       $message->message = $row[$prefix."message"];
-      $message->bugstatus = $this->bugstatus[intval($row[$prefix."statusId"])];
+      $message->BugStatus = $this->BugStatus[intval($row[$prefix."statusId"])];
       $message->comment = $row[$prefix."comment"];
       $message->assignee = new User();                  // TODO: load User
```
Comment (douira) [inline application/lib/MessageDbFactory.php]: I see this creates messages from database rows. Why doesn't it create MessageDbs? What is this class used for?

File: application/lib/MessageTypeDb.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+   
+   public function getAll() {
+		// Die Liste wird abgefragt
+		$query = $this->db->prepare("
+			SELECT
+				*
+			FROM message_type"
+		);
+		
+      return $this->getRows($query, new MessageTypeDbFactory());
```
Comment (douira) [inline application/lib/MessageTypeDb.php]: so this returns an array of messageTypes?

File: application/lib/Messagelist.php [hunk 6]
```diff
@@ -12,20 +12,20 @@ public function add(Message $msg) {
       $this->messages[] = $msg;
    }
```
Comment (douira) [inline application/lib/Messagelist.php]: Messagelist should maybe be renamed to MessageList?

Comment (douira) [inline application/lib/Messagelist.php]: Is this class a fancy array of message objects? What is the use case for the iterator methods below?

File: application/lib/ModuleDb.php [hunk 7]
```diff
@@ -3,6 +3,44 @@
 
 class ModuleDb extends \MVC\db\MySql {
```
Comment (douira) [inline application/lib/ModuleDb.php]: Essentially this is an interface for the module table. (as are the other *Dbs too...)

File: application/lib/StatusDb.php [hunk 8]
```diff
@@ -3,6 +3,29 @@
...
+            
+            PRIMARY KEY (device_id)
+         )
+         DEFAULT CHARACTER SET utf8 COLLATE utf8_bin
+      ");
+            
+      $query->execute();
+   }
+   
    public function getByIp(Ip $ip) {
```
Comment (douira) [inline application/lib/StatusDb.php]: Why does RoomDb have getAll() and StatusDb not?

File: application/templates/index_index.php [hunk 9]
```diff
@@ -4,29 +4,49 @@
...
+      <?php foreach($this->bugStatus as $bs): ?>
+         <label><?php echo $bs->name;?><input name="bs<?php echo $bs->id?>" type="checkbox" <?php if (in_array($bs->id, $this->bugStatusId)) echo "checked"?>></label>
+      <?php endforeach?>
+   </div>
+   <div class="form-group">
+      <input type="submit" name="filter" value="aktualisieren">
+   </div>
+</form>
 <div class='roomTable row'>
-   <table class='table-responsive table table-sm table-hover'>
```
Comment (douira) [inline application/templates/index_index.php]: I thought the hover effect was quite nice...

File: application/templates/index_index.php [hunk 10]
```diff
@@ -4,29 +4,49 @@
...
       </thead>
       <tbody>
-         <?php foreach ($this->devices as $device): ?>
-            <tr onclick='document.location = "/device?ip=<?php echo $device->ip->getInt();?>"'>
-               <td><?php echo $device->roomNr > 0 ? sprintf("%03d", $device->roomNr) : "unbekannt"; ?></td>
-               <?php foreach ($this->types as $type): ?>
-                  <td>
-                     <?php foreach ($this->bugstatus as $currBugstatus): ?>
-                        <?php $count = $device->messages->countBugstatus($currBugstatus, $type); ?>
-                        <span class='tag text-right <?php echo $count > 0 ? $currBugstatus->tagName : "tag-default"; ?>'><?php echo $count?></span>&nbsp;
```
Comment (douira) [inline application/templates/index_index.php]: I think I'll improve the spacing of the tags later.

File: application/controller/ControllerIndex.php [hunk 11]
```diff
@@ -8,18 +8,66 @@ function actionIndex(\MVC\RequestHttp $request) {
...
+            }
+         }
+      }
+      else {
+         $bugStatusId = array(1,2,3);                            // if filter not posted, default values
+         $messageTypeId = array(1,4,5,6);
+      }
       
-      $devices = $deviceDb->getWithMessages();
-      usort($devices, function($a,$b){return $a->roomNr - $b->roomNr;});      // sort by roomNr
```
Comment (tmshl) [inline application/controller/ControllerIndex.php]: good point. At least the room numbers need to be sorted. I will change the mysql-query soon.

File: application/lib/DeviceDbFactory.php (hunk 1 above)
Comment (tmshl) [inline application/lib/DeviceDbFactory.php]: This method is inherited from \MVC\db\DbFactory. It just fills the object (here $device) with the data from the database row (here $row). If there is a prefix, only prefixed columns are used and it ignores the prefix.

   ```
/**
    * Fills the given object with data from a database row.
    * @param $obj The object to be filled
    * @param array $row The database row. Needs to be an associative array which names are equal to attributes of the object (f.e. $row['foo'] --> $obj->foo)
    * @param string prefix Every column that is prefixed is processed. The prefix is not part of the attribute (f.e. $prefix='pr_' --> $row['pr_foo'] --> $obj->foo)
    * @param int $options NONE|CAMELCASE CAMELCASE transforms underlined keys to camelcased attributes (f.e. $row['foo_bar'] --> $obj->fooBar
    */
	protected function fill(&$obj, $row, $prefix="", $options=self::NONE) {	
```

File: application/lib/MessageDb.php (hunk 2 above)
Comment (tmshl) [inline application/lib/MessageDb.php]: Maybe in future. Now it does nothing. I don't know if it will be used.

File: application/lib/MessageDbFactory.php (hunk 3 above)
Comment (tmshl) [inline application/lib/MessageDbFactory.php]: No. getInstance is a method in Singleton classes. There is a discussion if Singleton is good practise but I consider it useful in cases when you need to ensure that there is only one instance of the class. In our program every *Db class inherits \MVC\db\Db, which is a Singleton. Because of this there will be only one database connection open. Not one for every *Db object.

[Here](http://www.phptherightway.com/pages/Design-Patterns.html) is more information about the design patterns used in this project. Especially Factory and Singleton.

File: application/lib/MessageDbFactory.php (hunk 4 above)
Comment (tmshl) [inline application/lib/MessageDbFactory.php]: A lot of classes that are connected to database tables are created with "factories". Consider we decide to change either the Message class or the message table in DB. Without factories we would have to review the whole code to ensure, that the data from the table is properly written into the object. If every instance of Message is created using the factory, we would just change the logic in the *Factory class. [Here](http://www.phptherightway.com/pages/Design-Patterns.html) is more information about the design patterns used in this project.

File: application/lib/Messagelist.php (hunk 6 above)
Comment (tmshl) [inline application/lib/Messagelist.php]: Actually this class was used by the old start page. I think it is useless now. But I didn't delete it because it could be of use later. And yes, it is just a fancy array. I thought that we often deal with lists of messages in our project, that it would be worth to have an easier access. In PHP every class, that implements the 'Iterator' interface can be iterated in foreach (...). The Iterator functions supply the functionality that is internally used by foreach and similar functions.

File: application/controller/ControllerIndex.php [hunk 12]
```diff
@@ -8,18 +8,66 @@ function actionIndex(\MVC\RequestHttp $request) {
...
+      // get information about message types, bug status
+      $response->messageTypes = MessageTypeDb::getInstance($dbConfig)->getAll();
+      $response->bugStatus = BugStatusDb::getInstance($dbConfig)->getAll();
 
-      // load all devices, which sent messages
-      $deviceDb = DeviceDb::getInstance(\MVC\MVC::getInstance()->data->db);
-      $deviceDb->provideTypes($response->types);
-      $deviceDb->provideBugstatus($response->bugstatus);
+      // setup filter
+      if ($request->has("filter")) {
```
Comment (douira) [inline application/controller/ControllerIndex.php]: Maybe the setting can be in a default easy state and changeable if needed so advanced and less advanced users can use it.

File: application/lib/MessageDb.php (hunk 2 above)
Comment (douira) [inline application/lib/MessageDb.php]: oh ok, there just seemed to be some uncertainty throughout the different classes concerning getAll() in some cases.

File: application/lib/StatusDb.php (hunk 8 above)
Comment (tmshl) [inline application/lib/StatusDb.php]: We need to get a list of all rooms in controller index, action index. A list of all deviceStatus is not needed yet an could be added later.
BTW: I forgot to change the class name an filename. It should be DeviceStatus. I will correct it soon.

File: application/templates/index_index.php (hunk 9 above)
Comment (tmshl) [inline application/templates/index_index.php]: ok. It just looked weird with the button within the table. Change the layout as You wish.

File: application/templates/index_index.php (hunk 10 above)
Comment (tmshl) [inline application/templates/index_index.php]: ok

File: application/lib/DeviceDbFactory.php (hunk 1 above)
Comment (douira) [inline application/lib/DeviceDbFactory.php]: Is there a reason that the factory methods are non-static? In what situation would one want to create a persistent factory object?

File: application/controller/ControllerIndex.php (hunk 12 above)
Comment (tmshl) [inline application/controller/ControllerIndex.php]: I agree.

File: application/lib/MessageDbFactory.php [hunk 13]
```diff
@@ -15,8 +15,8 @@ public function __construct(array $types, array $modules, array $bugstatus) {
          $this->modules[((string)$module->id)] = $module;
       }
       
-      foreach ($bugstatus as $bugst) {
-         $this->bugstatus[((string)$bugst->id)] = $bugst;
+      foreach ($BugStatus as $bugst) {
+         $this->BugStatus[((string)$bugst->id)] = $bugst;
```
Comment (douira) [inline application/lib/MessageDbFactory.php]: Why is this upper CamelCase now?

File: application/lib/DeviceDb.php [hunk 14]
```diff
@@ -2,95 +2,102 @@
...
-      $device->status = $statusDb->getByIp($ip);
-      
-      foreach($result as $row) {
-         $device->messages->add($msgFactory->create($row, "msg_"));
-      }
-      
-      return $device;
-   }
-   
-   protected function createMessageDbFactory() : MessageDbFactory {
```
Comment (douira) [inline application/lib/DeviceDb.php]: Method for creating a factory was confusing.

File: application/lib/DeviceDb.php [hunk 15]
```diff
@@ -2,95 +2,102 @@
...
-         // create new device, if unknown IP found
-         if (!isset($devices[$row["ip"]])) {
-            $ip = new Ip(intval($row["ip"]));
-            $roomNr = $roomDb->getByIp($ip);
-            $devices[$row["ip"]] = new Device($ip, $roomNr);
-         }
-         
-         // add current message
-         $devices[$row["ip"]]->messages->add($msgFactory->create($row));
+      $devices = $this->getRows($query, new DeviceDbFactory);
```
Comment (douira) [inline application/lib/DeviceDb.php]: Does object creation without parentheses just do the same as empty parentheses?

File: application/lib/MessageDbFactory.php (hunk 4 above)
Comment (douira) [inline application/lib/MessageDbFactory.php]: Ok, I read that. Very interesting. So we pass a factory to another method to be then used to create objects from database rows. (or just use the factory directly in fill() cases)

File: application/lib/MessageDbFactory.php (hunk 13 above)
Comment (tmshl) [inline application/lib/MessageDbFactory.php]: I don't know. Makes no sense. Please revert these changes yourself.

File: application/lib/DeviceDb.php (hunk 15 above)
Comment (tmshl) [inline application/lib/DeviceDb.php]: No. It is an error. Please correct it.

File: application/lib/DeviceDb.php (hunk 14 above)
Comment (tmshl) [inline application/lib/DeviceDb.php]: I agree. Some bad code, that is currently not used. We should delete this method later.
